### PR TITLE
TabPanel: Add prop to allow disabling of a tab button

### PR DESCRIPTION
**Pull request by joshuatf**
## What?
Adds a prop to allow disabling of a given tab.

## Why?
In WooCommerce we plan to disable tabs depending on a product configuration.  While we could hide the disabled tabs from view to solve this, it's helpful for users to see disabled tabs and hover them for hints as to why certain tabs are disabled.

## How?
Adds a prop to disable the tab.  Alternatively we could pass `...rest` to the tab button to allow passing of any props, but I'm not sure if this is wanted and would not have the advantages of TS.

## Testing Instructions
1. Run `npm run test:unit tab-panel ` and make sure all tests pass
2. Optionally create a tab panel with a a disabled prop
```
			<TabPanel
				tabs={ [
					{
						name: 'delta',
						title: 'Delta',
						className: 'delta-class',
						disabled: true,
					},
				] }
				children={ () => undefined }
				onSelect={ () => undefined }
			/>
```
3. Make sure the tab is disabled

## Screenshots or screencast 
Screen Shot 2022-12-12 at 10 22 17 AM

### CODE REVIEW

**Review by alexstine — COMMENTED** [review]
I think code looks good. I just need a little more time to A11Y check this. Passing `disabled` to `Button` component is okay since we are handling that for screen readers, but not sure how it behaves down below in `NavigableMenu`.

File: packages/components/src/tab-panel/types.ts
Comment (mirka) [inline packages/components/src/tab-panel/types.ts]: I think `disabled` (and `icon`, too) should be defined in the `Tab` type first and foremost, as they are now part of the standard `Tab` interface. And then they can be `Pick`ed into `TabButtonProps`.

For context, the `Record< any, any >` part of the `Tab` type is just a catch-all intended for consumers who want to pass arbitrary data down to the tabpanel render function. Personally I don't love this API design since it isn't necessary (i.e. consumers should be able to augment whatever data they want as long as they know the tab id), but I think we have to continue supporting it for back compat.

File: packages/components/src/tab-panel/test/index.tsx
Comment (mirka) [inline packages/components/src/tab-panel/test/index.tsx]: Thanks for adding the test ✨

Another thing I think we need to handle and test for is when the `selected` tab happens to be disabled. For example, on the initial load of a tab list like this:

```js
[
		{
			name: 'tab1',
			title: 'Tab 1',
			disabled: true,
		},
		{
			name: 'tab2',
			title: 'Tab 2',
		},
],
```

File: packages/components/src/tab-panel/index.tsx
Comment (mirka) [inline packages/components/src/tab-panel/index.tsx]: I feel like it might be better if the button were still focusable even when disabled. Especially since you mention in the PR description that it would be helpful for users to hover tabs for hints on why they're disabled. What do you think @alexstine?

There is a `__experimentalIsFocusable` prop we can enable on `Button` to make this happen.

Either way, it would be great if we could add an explicit keyboard navigation test for whichever behavior we decide on. Normally it's sufficient to just test for `toBeDisabled()`, but in this case there is a `NavigableMenu` layer that converts the usual tab-key-based navigation to be arrow-key-based, so I think it's worth testing that integration point.

File: packages/components/src/tab-panel/index.tsx
Comment (mirka) [inline packages/components/src/tab-panel/index.tsx]: Nit: No strong opinion, but the prop passing is getting a little verbose here so it might be a good time to simplify:

```jsx
const TabButton = ( { tabId, selected, ...rest }: TabButtonProps ) => (
	<Button
		role="tab"
		tabIndex={ selected ? null : -1 }
		aria-selected={ selected }
		id={ tabId }
		{ ...rest }
	/>
);
```

**Review by mirka — COMMENTED** [review]
A few details to work out, but this looks good to me as well 👍 Thanks!

Two more things from my end:

- We need to add a [changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) before merging.
- I think it would be helpful to add a Storybook story for this `disabled` use case, since it's a bit hard to discover at the moment. It would also be easier to catch styling bugs moving forward.

File: packages/components/src/tab-panel/index.tsx
Comment (alexstine) [reply packages/components/src/tab-panel/index.tsx]: That prop is set to true by default I think so should not be an issue. Need to test to confirm.

File: packages/components/src/tab-panel/types.ts
Comment (joshuatf) [reply packages/components/src/tab-panel/types.ts]: Added in https://github.com/WordPress/gutenberg/pull/46471/commits/3b0ea90b1014388662a069b25f112639bf797da4.

File: packages/components/src/tab-panel/test/index.tsx
Comment (joshuatf) [reply packages/components/src/tab-panel/test/index.tsx]: Adding this for selected and initially selected tabs in https://github.com/WordPress/gutenberg/pull/46471/commits/529683594274bbbb9cb6682a4bd5e91b507f698d.

File: packages/components/src/tab-panel/index.tsx
Comment (joshuatf) [reply packages/components/src/tab-panel/index.tsx]: I ended up having to use `__experimentalIsFocusable` in order to focus the disabled button.  Added in https://github.com/WordPress/gutenberg/pull/46471/commits/662bfd7d97e099bf7de425b8a887f35ced0166a4.

**Comment by joshuatf** [comment]
Thanks for the feedback @alexstine and @mirka.  I added some additional logic for updating selection when tabs are disabled as well as additional tests around this logic.

This is now visible in Storybook which demonstrates the initial tab selection, but you can also verify that selection updates to the first enabled tab if the current tab is disabled by editing the raw `tabs` value in Storybook.  E.g., select "Tab 3" and then update the prop data to:

```
[
  {
    "name": "tab1",
    "title": "Tab 1"
  },
  {
    "name": "tab2",
    "title": "Tab 2"
  },
  {
    "name": "tab3",
    "title": "Tab 3",
    "disabled": true
  }
]
```

Please let me know if everything looks good now- thanks!

**Comment by joshuatf** [comment]
Hi @mirka @alexstine - could you give this PR another look when you have time?  Thanks!

**Review by mirka — APPROVED** [review]
Beautiful, thank you ✨ Code is clean, good fallback logic, and good test coverage. Accessibility looks good to me as well — markup and keyboard nav works as expected.

I think we're good to merge, unless @alexstine has any other concerns or would like to block it until he has some time to test.

**Review by alexstine — APPROVED** [review]
@joshuatf Sorry for the delay, I have been a little under the weather lately.

Looks good.

**Comment by mirka** [comment]
Almost forgot, let's update the [component readme](https://github.com/WordPress/gutenberg/blob/66091754995f286d8e6fd688bd88ab6c19643a5a/packages/components/src/tab-panel/README.md?plain=1#L116) before we merge.

**Comment by joshuatf** [comment]
Thank you both for the reviews, @alexstine and @mirka!  Updated the readme in https://github.com/WordPress/gutenberg/pull/46471/commits/b99186265db8aa73bf81b7bc9d80535d282e968b.

File: packages/components/src/tab-panel/index.tsx
Comment (ciampo) [inline packages/components/src/tab-panel/index.tsx]: Nitty, but — right now, the code makes it look like the `handleTabSelection` function can be called twice during the same effect.

Should we rewrite the logic slightly (e.g. `else if`) to make it clear that these two code branches should be exclusive?

Comment (joshuatf) [reply packages/components/src/tab-panel/index.tsx]: Sure!  Done in https://github.com/WordPress/gutenberg/pull/46471/commits/c64730cc15148882be85a50c5536de3261138152.

**Review by ciampo — APPROVED** [review]
Thank you for working on this! 🎉

**Comment by scruffian** [comment]
I think this might have caused this issue: https://github.com/WordPress/gutenberg/issues/47079

**Comment by talldan** [comment]
> I think this might have caused this issue: https://github.com/WordPress/gutenberg/issues/47079

It seems like there are two causes:
1. The Navigation block's List View tab isn't declared on first render, it only becomes present in the `tabs` array on later renders.
2. The TabPanel no longer supports lazily adding initial tabs after this PR, it now always falls back to the first enabled tab in the `tabs` array.

I did try looking to see if this could be fixed by always rendering the tab navigation block, but I couldn't find the source of the problem. It could be worth some further investigation though.

I've put together a fix for the `TabPanel` component - https://github.com/WordPress/gutenberg/pull/47100.

**Pull request merged**
